### PR TITLE
Make plots resize with windows better

### LINKED ISSUE #35014: X Axis Title of the graphs hides when we change size (second fix)
**Original reporter:** A new issue for the same bug as reported in #34049 as the original fix needed to be reverted as it caused other layouting issues. See the referenced issue for more details. We reproduce the initial bug content here.

This was raised as a feature request in a support email from Leide Cavalcanti that was sent on 18th May 2022 along with some beta testing feedback on the 6.4 release candidate:

The X Axis Title of the graphs hides when we change size (please see attachment)

The issue is #2 in the attached document

[Mantid_graphs.pdf](https://github.com/mantidproject/mantid/files/8891535/Mantid_graphs.pdf)

### PR DESCRIPTION
**Pull request by jhaigh0**
**Description of work.**

This PR fixes a bug where plot axes labels could be hidden outside the window boundaries after resizing. It also aims to make all plot types resize in a more predictable and nicer way. 

An important part of this was giving plots a tight layout; this has resulted in warnings being printed to the console when a window is resized to be very small. Since these warnings are annoying, I tried to filter them out, but couldn't find a solution that worked. However, you have to shrink the plot to an unusable size to make them appear, so I think it is okay overall.

**To test:**

To test these changes, it will be useful to have another version of Mantid open to see the differences.

 - Load `164199.nxs` (from the training course data)
 - Right-clicking on the loaded workspace, plot a....
 	- [x] Spectrum
 	- [x] Spectrum with errors
 	- [x] Spectrum range tiled plot
 	- [x] Spectrum range waterfall plot
 	- [x] Super Plot
 	- [x] Super Plot with errors
 	- [x] Colorfill
 	- [x] Surface
 	- [x] Wireframe
 	- [x] Contour
 - And for each of these resize the window to a few different **reasonable** sizes and check that
	1. No figure objects are colliding
	2. No parts of the figure are being hidden outside the window boundaries
	3. Moving the legend and editing axes names and ranges doesn't break anything
 - Now load `164198.nxs` and `164200.nxs`
 - Group the three workspaces together (highlight all and click Group)
 - Right-clicking on the new group, plot all the different types of plots again and make the same checks

Finally, test with a some other graphing scenarios that you think might be challenging.

Fixes #35014

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Are the release notes saved in a separate file, using Issue or PR number for file name and in the correct location?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: Framework/PythonInterface/mantid/plots/plotfunctions.py [hunk 1]
```diff
@@ -225,8 +226,11 @@ def plot(
 
     show_legend = "on" == ConfigService.getString("plots.ShowLegend").lower()
     for ax in axes:
-        if ax.get_legend() is not None:
-            ax.get_legend().set_visible(show_legend)
+        legend = ax.get_legend()
+        if legend is not None:
+            legend.set_visible(show_legend)
+            # Stop legend messing with the tight layout
```
Comment (sf1919) [inline Framework/PythonInterface/mantid/plots/plotfunctions.py:232]:
```suggestion
            # Stop legend interfering with the tight layout
```

File: Framework/PythonInterface/mantid/plots/plotfunctions.py [hunk 2]
```diff
@@ -324,15 +330,43 @@ def create_subplots(nplots, fig=None):
...
-    for i in range(1, nplots):
-        axes[i] = fig.add_subplot(gs[i // ncols, i % ncols], projection=PROJECTION)
-    axes = axes.reshape(nrows, ncols)
+    cbar_axis = None
+    fig.set_layout_engine(layout="tight")
+
+    if add_cbar_axis:
+        # The right most column of the GridSpec is made a SubGridSpec to facilitate the colour bar
+        # This is done so that the colour bar can have a close spacing to the right most column
+        # Keeping the colour bar in the GridSpec rather than alongside it means it behaves much nicer
```
Comment (sf1919) [inline Framework/PythonInterface/mantid/plots/plotfunctions.py:340]: Could this be reworded slightly? It seems a bit subjective and not clear what behaviour it is actually doing.

**Review by sf1919 — CHANGES_REQUESTED** [review]
This works very well and it's useful that when the labels etc are cut off you do get an error message - `Tight layout not applied. The left and right margins cannot be made large enough to accommodate all axes decorations.`

I've asked for a couple of changes to the comments just to make them clearer.

Should this have test(s)?

**Comment by jhaigh0** [comment]
Yes, I reckon I could've put a test in for the creating subplots colour bar change; I'll add that.

**Review by sf1919 — APPROVED** [review]
Thank you for putting the tests in. I think they're a good addition and should hopefully catch anything like this in future. 

This one is now good to go.

**Pull request merged**